### ISSUE TITLE
sync: negative WaitGroup counter

### ISSUE THREAD

**Issue by jgawor** [cloudfoundry/cli#1019]
cf push sometimes fails with:

```
panic: sync: negative WaitGroup counter

goroutine 1359 [running]:
panic(0xb08e60, 0xc4204ce730)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
sync.(*WaitGroup).Add(0xc42070e930, 0xffffffffffffffff)
	/usr/local/go/src/sync/waitgroup.go:75 +0x1fe
sync.(*WaitGroup).Done(0xc42070e930)
	/usr/local/go/src/sync/waitgroup.go:100 +0x34
code.cloudfoundry.org/cli/cf/commands/application.(*Start).TailStagingLogs.func1()
	/tmp/build/80754af9/gopath/src/code.cloudfoundry.org/cli/cf/commands/application/start.go:253 +0xdc
code.cloudfoundry.org/cli/cf/api/logs.(*NoaaLogsRepository).TailLogsFor.func1()
	/tmp/build/80754af9/gopath/src/code.cloudfoundry.org/cli/cf/api/logs/noaa_logs_repository.go:74 +0x3d
code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer.(*Consumer).retryAction.func1()
	/tmp/build/80754af9/gopath/src/code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer/async.go:274 +0x78
code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer.(*Consumer).establishWebsocketConnection(0xc4203d8500, 0xc4201d04c0, 0x31, 0xc42070c700, 0x326, 0x93713, 0x2, 0xc420e01d80)
	/tmp/build/80754af9/gopath/src/code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer/async.go:370 +0xbc
code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer.(*Consumer).listenAction.func1(0xc4203d85f8, 0xdf8475800, 0x77359400)
	/tmp/build/80754af9/gopath/src/code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer/async.go:257 +0xaf
code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer.(*Consumer).retryAction(0xc4203d8500, 0xc4201d0500, 0xc4203ea720)
	/tmp/build/80754af9/gopath/src/code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer/async.go:279 +0x158
code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer.(*Consumer).streamAppDataTo(0xc4203d8500, 0xc420e128e0, 0xc42029ede0, 0x24, 0xc42070c700, 0x326, 0xc42070e970, 0xc4203ea720, 0x82df01)
	/tmp/build/80754af9/gopath/src/code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer/async.go:188 +0x14a
code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer.(*Consumer).tailingLogs.func2(0xc4203ea720, 0xc4203ea6c0, 0xc4203d8500, 0xc420e128e0, 0xc42029ede0, 0x24, 0xc42070c700, 0x326, 0xc42070e970, 0x451e01)
	/tmp/build/80754af9/gopath/src/code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer/async.go:163 +0xde
created by code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer.(*Consumer).tailingLogs
	/tmp/build/80754af9/gopath/src/code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer/async.go:164 +0x147
```

The command executed looks like the following:
```
$ cf push appName -f manifest.yml -b buildpackName 
```

CF CLI info:
```
$ cf --version
cf version 6.22.2+a95e24c-2016-10-27
$ uname -a
Linux XXX 3.13.0-101-generic #148-Ubuntu SMP Thu Oct 20 22:08:32 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by XenoPhex** [comment]
The diego team found a similar bug:
```
cf push smoke-test-f436f86f-917a-4bf2-77fe-fa62735757b5 -p dora -i 1 -b ruby_buildpack 
Creating app smoke-test-f436f86f-917a-4bf2-77fe-fa62735757b5 in org org-98caaf6a-8ff4-4bf8-600e-212aef4b6397 / space space-3e2a29c6-b5ea-4ff7-5c78-e79b4279554d as admin...
OK
Creating route smoke-test-f436f86f-917a-4bf2-77fe-fa62735757b5.dusts-mysql.diego-ci.cf-app.com...
OK
Binding smoke-test-f436f86f-917a-4bf2-77fe-fa62735757b5.dusts-mysql.diego-ci.cf-app.com to smoke-test-f436f86f-917a-4bf2-77fe-fa62735757b5...
OK
Uploading smoke-test-f436f86f-917a-4bf2-77fe-fa62735757b5...
Uploading app files from: /tmp/build/c933e80d/diego-release/src/code.cloudfoundry.org/diego-upgrade-stability-tests/dora
Uploading 137.8K, 38 files
Done uploading               
OK
Warning: error tailing logs
Timed out waiting for connection to Loggregator (wss://doppler.dusts-mysql.diego-ci.cf-app.com:4443).
Starting app smoke-test-f436f86f-917a-4bf2-77fe-fa62735757b5 in org org-98caaf6a-8ff4-4bf8-600e-212aef4b6397 / space space-3e2a29c6-b5ea-4ff7-5c78-e79b4279554d as admin...
panic: sync: negative WaitGroup counter
goroutine 981 [running]:
panic(0xb08e60, 0xc420396390)
    /usr/local/go/src/runtime/panic.go:500 +0x1a1
sync.(*WaitGroup).Add(0xc4204626e0, 0xffffffffffffffff)
    /usr/local/go/src/sync/waitgroup.go:75 +0x1fe
sync.(*WaitGroup).Done(0xc4204626e0)
    /usr/local/go/src/sync/waitgroup.go:100 +0x34
code.cloudfoundry.org/cli/cf/commands/application.(*Start).TailStagingLogs.func1()
    /tmp/build/80754af9/gopath/src/code.cloudfoundry.org/cli/cf/commands/application/start.go:253 +0xdc
code.cloudfoundry.org/cli/cf/api/logs.(*NoaaLogsRepository).TailLogsFor.func1()
    /tmp/build/80754af9/gopath/src/code.cloudfoundry.org/cli/cf/api/logs/noaa_logs_repository.go:74 +0x3d
code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer.(*Consumer).retryAction.func1()
    /tmp/build/80754af9/gopath/src/code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer/async.go:274 +0x78
code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer.(*Consumer).establishWebsocketConnection(0xc42008cd00, 0xc4200fe600, 0x31, 0xc42062b900, 0x4a5, 0x127a13, 0x2, 0xc420955d70)
    /tmp/build/80754af9/gopath/src/code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer/async.go:370 +0xbc
code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer.(*Consumer).listenAction.func1(0xc42008cdf8, 0xdf8475800, 0x1dcd65000)
    /tmp/build/80754af9/gopath/src/code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer/async.go:257 +0xaf
code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer.(*Consumer).retryAction(0xc42008cd00, 0xc4200fe640, 0xc4202e1860)
    /tmp/build/80754af9/gopath/src/code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer/async.go:279 +0x158
code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer.(*Consumer).streamAppDataTo(0xc42008cd00, 0xc420437f40, 0xc420202f90, 0x24, 0xc42062b900, 0x4a5, 0xc420462760, 0xc4202e1860, 0x180001)
    /tmp/build/80754af9/gopath/src/code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer/async.go:188 +0x14a
code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer.(*Consumer).tailingLogs.func2(0xc4202e1860, 0xc4202e17a0, 0xc42008cd00, 0xc420437f40, 0xc420202f90, 0x24, 0xc42062b900, 0x4a5, 0xc420462760, 0x451e01)
    /tmp/build/80754af9/gopath/src/code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer/async.go:163 +0xde
created by code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer.(*Consumer).tailingLogs
    /tmp/build/80754af9/gopath/src/code.cloudfoundry.org/cli/vendor/github.com/cloudfoundry/noaa/consumer/async.go:164 +0x147
STEP: Test Complete, AfterEach Beginning
STEP: exiting continuous test poller (0 failed curl requests / 10 attempted curl requests)
```

**Comment by dkoper** [comment]
This issue has now been resolved in cf CLI 6.23.0.
https://github.com/cloudfoundry/cli#downloads

Regards,
Dies Koper
CF CLI PM
